cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -2145,7 +2166,10 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
           child_path_ss << bucket << kSep << child_key;
           child_key = child_path_ss.str();
           if (obj.GetSize() > 0 || !had_trailing_slash) {
-            // We found a real file
+            // We found a real file.
+            // XXX Ideally, for 0-sized files we would also check the Content-Type
+            // against kAwsDirectoryContentType, but ListObjectsV2 does not give
+            // that information.
```

Comment:
As I said above, trying to support this gets ugly very quickly. The explosion of conditions to check when validating operations and extra requests everywhere.

Reply:
I checked the issue and I think supporting directories with trailing-slash-less keys + requiring that files never have a trailing-slash is enough to solve the issue. And won't create more issues.

Reply:
I agree it's not very pretty, but if it solves a real issue for users, then we should probably do it IMHO.

Reply:
By ugly I mean it creates bugs: actions on the filesystem become undefined or with very complicated semantics.

Reply:
For example:

s3fs doesn't support `Move` yet I think, but if it were to support, the semantics of the path validations take the trailing `/` in `dest` to mean `src` should also be a directory. It's OK to pass `/`-less to `Move` that refer to directories, but it's not OK to pass paths with trailing slashes that refer to files.

This prevents accidental overrides of files by `Move(src, dest)` because directories can only be overriden if they are empty.

Reply:
`GetFileInfo` has the following post-condition:

If path has trailing slash, then result is either `FileType::kDirectory` or `FileType::kNotFound` [1]. You would break that by returning `kFile` for paths containing trailing slashes.

> I don't think we plan to support Move because, last I looked, S3 didn't support a native move operation.

Technically, it doesn't support any directory operation, but the file system implementation is supposed to be a layer that implements filesytem semantics on top of it. Even a filesystem `stat` can't be mapped directly to S3 operations.

[1] https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/localfs.cc#L169

Reply:
> I agree it's not very pretty, but if it solves a real issue for users, then we should probably do it IMHO.

The restriction I'm asking for doesn't prevent the issue to be fixed. We can support directories on paths not ending with a `/`. I'm asking to not expand the scope of the fix to also allow files on paths with trailing `/`.

Reply:
But again, the PR doesn't change anything about this. We could make our heuristics stricter but that would be beside the scope of this PR.

Reply:
You're right. All the calls to `IsDirectory` were accepting paths that end with `/` as files. 🤔 

Can you at least add a comment on the file size check in `IsDirectory`?

Reply:
I'll add a comment, yes.